nautobot/core/ui/object_detail.py

```diff
@@ -0,0 +1,567 @@
+                        ),
+                        # TODO add advanced_ui custom fields and relationships
+                    ),
+                ),
+                LayoutFullWidth(),
+            )
+        )
+
+
+class _ObjectDetailContactsTab(Tab):
+    """Built-in class for a Tab displaying information about contact/team associations."""
+
+    tab_id = "contacts"
+
+    def __init__(self):
+        super().__init__(
+            layouts=(
+                LayoutFullWidth(
+                    panels=(
+                        # TODO
+                        TemplatePanel(label="Contact Associations", template_string="Contact Associations Table"),
+                    ),
+                ),
+            ),
+        )
+
+    def should_render(self, context):
+        return context["object"].is_contact_associable_model
+
+    @property
+    def tab_label(self):
+        return format_html(
+            "Contacts {}",
+            render_to_string(
+                "utilities/templatetags/badge.html", badge(self.context["object"].associated_contacts.count())
+            ),
+        )
+
+
+class _ObjectDetailGroupsTab(Tab):
+    """Built-in class for a Tab displaying information about associated dynamic groups."""
+
+    tab_id = "dynamic_groups"
+
+    def __init__(self):
+        super().__init__(
+            layouts=(
+                LayoutFullWidth(
+                    panels=(
+                        # TODO
+                        TemplatePanel(label="Dynamic Groups", template_string="Dynamic Groups Table"),
+                    ),
+                ),
+            ),
+        )
+
+    def should_render(self, context):
+        return (
+            context["object"].is_dynamic_group_associable_model
+            and context["request"].user.has_perm("extras.view_dynamicgroup")
+            and context["object"].dynamic_groups.exists()
+        )
+
+    @property
+    def tab_label(self):
+        return format_html(
+            "Dynamic Groups {}",
+            render_to_string("utilities/templatetags/badge.html", badge(self.context["object"].dynamic_groups.count())),
+        )
+
+
+class _ObjectDetailMetadataTab(Tab):
+    """Built-in class for a Tab displaying information about associated object metadata."""
+
+    tab_id = "object_metadata"
+
+    def __init__(self):
+        super().__init__(
+            layouts=(
+                LayoutFullWidth(
+                    panels=(
+                        # TODO
+                        TemplatePanel(label="Object Metadata", template_string="Object Metadata Table"),
+                    ),
+                ),
+            ),
+        )
+
+    def should_render(self, context):
+        return (
+            context["object"].is_metadata_associable_model
+            and context["request"].user.has_perm("extras.view_objectmetadata")
+            and context["object"].associated_object_metadata.exists()
+        )
```

Comment:
nitpick: should we swap the orders here?
```suggestion
        return (
            context["object"].associated_object_metadata.exists()
            and context["request"].user.has_perm("extras.view_objectmetadata")
            and context["object"].is_metadata_associable_model
        )
```

Reply:
On second thought, do we need `context["object"].is_metadata_associable_model` here? `context["object"].associated_object_metadata.exists()` should just be enough right?

Reply:
The `is_metadata_associable_model` check is cheap since it doesn't require a database hit, unlike the other two, so I think it makes sense as a quick-escape.